### PR TITLE
WIP: Use Polly to mock HTTP requests in tests

### PR DESCRIPTION
**Pull request by mcmire**
**tl;dr:** I've introduced [Polly][1] as a replacement for Nock as I
believe it will lead to more readable and authentic tests. A few things:

* This is a proof of concept – I've only added it to
  CollectiblesController to start out with.
* All I've done is replace Nock requests with Polly requests. Ideally,
  we wouldn't need to mock anything manually at all and let Polly do its
  thing. Unfortunately, some HTTP requests that we are mocking no longer
  work, so we have to keep mocking for now. That can be refactored in a
  future commit.
* In the future, we should investigate using Ganache instead of hitting
  mainnet and exposing our Infura project ID in tests.

---

Currently there are a couple of problems with our tests as it relates to
network requests:

1. We do not prevent network requests from being made. There are
   some tests which use Nock to mock requests, but not all requests are
   being mocked, and we are unaware which ones are actually going
   through.
2. Nock, although a popular package, is rather cumbersome to use, and
   makes tests difficult to understand and follow.

To address these problems, this commit replaces Nock with Polly. Polly
is a package developed by Netflix that intercepts all HTTP requests and
automatically captures snapshots (recordings) of these requests as they
are being made. These recordings are then used in successive test runs.
You can also manually mock requests if you want more manual control.

## Why introduce another package? Why is Nock not sufficient?

There are a few problems with Nock that make it difficult to work with:

1. Mocks take up space in your test, so you will probably find yourself
   extending a collection of mocks you've assembled at the top of the
   file. Once you have enough, who knows which mocks serve which tests?
2. Once you've added a mock to your test, that mock stays forever. If
   the API changes in the future, or stops working altogether, you will
   never know until you remove that mock.
3. If you turn `nock.disableNetConnect()` on, and you haven't mocked a
   request, [it will throw a NetConnectNotAllowedError][2]. This works
   in most cases, but what happens if you have a `try`/`catch` block
   around the request you're making and you are swallowing the error?
   Then your test may fail for no apparent reason. And how do you know
   which request you need to mock? Nock has a way to peek behind the
   curtain and log the requests that are being made by running
   `DEBUG=nock.* yarn jest ...`, but this output is seriously difficult
   to read:

   ```
   2022-03-23T03:50:09.033Z nock.scope:api.opensea.io query matching skipped
   2022-03-23T03:50:09.033Z nock.scope:api.opensea.io matching [https://api.opensea.io:443/api/v1/asset/0x495f947276749Ce646f68AC8c248420045cb7b5e/40815311521795738946686668571398122012172359753720345430028676522525371400193](https://api.opensea.io/api/v1/asset/0x495f947276749Ce646f68AC8c248420045cb7b5e/40815311521795738946686668571398122012172359753720345430028676522525371400193) to GET
   [https://api.opensea.io:443/api/v1/asset_contract/0x2aEa4Add166EBf38b63d09a75dE1a7b94Aa24163](https://api.opensea.io/api/v1/asset_contract/0x2aEa4Add166EBf38b63d09a75dE1a7b94Aa24163): false
   2022-03-23T03:50:09.033Z nock.scope:api.opensea.io attempting match {"protocol":"https:","slashes":true,"auth":null,"host":"api.opensea.io:443","port":443,"hostname":"api.opensea.io","hash":null,"search":null,"query":null,"pathname":"/api/v1/asset/0x495f947276749Ce646f68AC8c248420045cb7b5e/40815311521795738946686668571398122012172359753720345430028676522525371400193","path":"/api/v1/asset/0x495f947276749Ce646f68AC8c248420045cb7b5e/40815311521795738946686668571398122012172359753720345430028676522525371400193","href":"[https://api.opensea.io/api/v1/asset/0x495f947276749Ce646f68AC8c248420045cb7b5e/40815311521795738946686668571398122012172359753720345430028676522525371400193","method":"GET","headers":{"accept":["*/*"],"user-agent":["node-fetch/1.0](https://api.opensea.io/api/v1/asset/0x495f947276749Ce646f68AC8c248420045cb7b5e/40815311521795738946686668571398122012172359753720345430028676522525371400193%22,%22method%22:%22GET%22,%22headers%22:%7B%22accept%22:[%22*/*%22],%22user-agent%22:[%22node-fetch/1.0) (+[https://github.com/bitinn/node-fetch)"],"accept-encoding":["gzip,deflate"],"connection":["close"]},"proto":"https](https://github.com/bitinn/node-fetch)%22],%22accept-encoding%22:[%22gzip,deflate%22],%22connection%22:[%22close%22]%7D,%22proto%22:%22https)"}, body = ""
   ```

Polly removes all of these problems:

1. Because Polly records request snapshots, you don't usually have to
   worry about mocking requests manually. These recordings are kept in
   files which are categorized by the tests which made the requests.
2. Because Polly actually hits the API you're mocking, you can work with
   real data. You can instruct Polly to expire these recordings after a
   designated timeframe to guarantee that your code doesn't break if
   the API ceases to work in the way you expect.
3. Polly will also print a message when a request is made that it
   doesn't recognize without needing to run an extra command:

    ```
    Errored ➞ POST https://mainnet.infura.io/v3/ad3a368836ff4596becc3be8e2f137ac PollyError: [Polly] [adapter:node-http] Recording for the following request is not found and `recordIfMissing` is `false`.
        {
          "url": "https://mainnet.infura.io/v3/ad3a368836ff4596becc3be8e2f137ac",
          "method": "POST",
          "headers": {
            "content-type": "application/json",
            "connection": "keep-alive",
            "host": "mainnet.infura.io",
            "user-agent": "Mozilla/5.0 (Darwin x64) node.js/12.22.6 v8/7.8.279.23-node.56",
            "content-length": "201"
          },
          "body": "{\"jsonrpc\":\"2.0\",\"id\":30,\"method\":\"eth_call\",\"params\":[{\"to\":\"0x18E8E76aeB9E2d9FA2A2b88DD9CF3C8ED45c3660\",\"data\":\"0x0e89341c0000000000000000000000000000000000000000000000000000000000000024\"},\"latest\"]}",
          "recordingName": "CollectiblesController/updateCollectibleFavoriteStatus/should keep the favorite status as false after updating metadata",
          "id": "53192eab94ddedfb300b625b1cb79843",
    ...
    ```

## What about Nock Back? Doesn't this do the same thing?

It's true that Nock contains a feature to capture requests called [Nock
Back][3]. However, this is cumbersome, too:

1. You have to wrap your test in a `nockBack` call.
2. You have to name your recording files.
3. You have to call `nockDone()` in your test.

There is a package called `jest-nock-back` that fixes these issues and
makes using Nock Back very easy. However, it isn't maintained (last
release was 3 years ago) and is likely incompatible with the newest
version of Jest.

## What about requests that irreversibly change data?

Some requests make irreversible changes to resources, such as
transferring a token from one account to another.

To handle these types of requests, Polly still lets you manually mock
requests just like Nock. We've configured Polly to not record a request
it doesn't recognize by default, so in this case you'll get a message
when running the test that makes such a request. From there you can make
a decision about how you'd like to mock this request — whether you want
Polly to record the request or whether you'd like to manually mock it.

That said, if the request in question involves a blockchain network,
instead of mocking the request, we might investigate whether it's
possible to use Ganache to perform the irreversible action rather than
actually hitting mainnet or some other network.

## What are the tradeoffs for Polly vs. Nock?

* Because Polly automatically creates request mocks and these mocks are
  no longer contained in tests, developers will need to be educated
  about Polly (Polly is not as popular as Nock) and remember that it
  exists whenever updating tests.
* If a recording is no longer needed or needs to be updated, the
  associated file holding the recorded information about the request
  needs to be deleted. Finding this file is not a great experience.
* Nock has virtually zero initialization code. Polly's initialization
  code is surprisingly complicated (although that is largely hidden, so
  no one should have to mess with it).

[1]: https://netflix.github.io/pollyjs/#/README
[2]: https://github.com/nock/nock#enabledisable-real-http-requests
[3]: https://github.com/nock/nock#nock-back

### CODE REVIEW

File: src/assets/CollectiblesController.test.ts [hunk 1]
```diff
@@ -561,10 +627,10 @@ describe('CollectiblesController', () => {
       ).toStrictEqual([
         {
           address: ERC721_KUDOSADDRESS,
-          description: 'Kudos Description',
```
Comment (mcmire) [inline src/assets/CollectiblesController.test.ts:564]: I am not 100% sure why these tests changed after I replaced the Nock mocks. I noticed that some requests were being called for real and I attempted to mock those. It's possible I missed something. If this were not a draft PR, I would correct this, but for the time being, I just tried to get the tests passing.

File: tests/setupTestsAfterEnv.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import path from 'path';
+import { setupPolly } from 'setup-polly-jest';
+import NodeHttpAdapter from '@pollyjs/adapter-node-http';
+import FSPersister from '@pollyjs/persister-fs';
+
+global.pollyContext = setupPolly({
```
Comment (mcmire) [inline tests/setupTestsAfterEnv.ts:6]: This is a bit of a hack to set up Polly globally and make it accessible in any test. It is a variation on these instructions: https://netflix.github.io/pollyjs/#/test-frameworks/jest-jasmine

File: scripts/check-for-expired-polly-recordings.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env node
```
Comment (mcmire) [inline scripts/check-for-expired-polly-recordings.js:1]: Polly has an [`expiresIn` setting](https://netflix.github.io/pollyjs/#/configuration?id=expiresin), which causes the recordings Polly has captured to automatically expire after a set time. Since we set `recordIfMissing: false`, this will cause tests with expired recordings to fail tests, which may force developers to recapture recordings for tests that they've never touched. For this reason, I've opted to write this script to check for expired recordings instead of using `expiresIn`. That way we could run this on CI or locally or wherever and it wouldn't interfere with development.

File: tests/setupTestsAfterEnv.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  adapters: [NodeHttpAdapter],
+  persister: FSPersister,
+  persisterOptions: {
+    fs: {
+      recordingsDir: path.resolve(__dirname, '__recordings__'),
+    },
+  },
+  recordIfMissing: false,
```
Comment (mcmire) [inline tests/setupTestsAfterEnv.ts:14]: Usually Polly records requests by default. That is fine if you are adding Polly to a fresh project, but since we're adding Polly to an existing project we need to know whether there are requests being made in tests that we forgot to mock using Nock. This allows us to do this — we will see a warning if Polly sees a request it doesn't recognize.

**Comment by mcmire** [comment]
I've made a couple of changes to this PR (notably a script to look for outdated recordings/snapshots) which means that we _could_ conceivably use Nock Back instead of Polly. I'll try some more things out tomorrow.

File: .gitattributes [hunk 5]
```diff
@@ -0,0 +1,2 @@
+# Collapse changes to Polly recordings in PRs.
+tests/__recordings__/** linguist-generated
```
Comment (Gudahtt) [inline .gitattributes:2]: ... huh. Part of our typical repo setup s a `.gitattributes` file that sets `* text=auto`, to normalize line endings. We also instruct linguist not to collapse lockfile changes.

I guess we forgot it here.

File: tests/setupTestsAfterEnv.ts (hunk 4 above)
Comment (Gudahtt) [reply tests/setupTestsAfterEnv.ts:14]: Hmm. I was expecting there to be some way to throw if a record is missing. Otherwise we might miss these warnings creeping in, and we might not see tests fail either for the same reason you mentioned for `disableNetConnect`.

Comment (Gudahtt) [reply tests/setupTestsAfterEnv.ts:14]: Honestly this is the main sticking point for me at the moment. I'd feel better knowing that it was impossible for it to hit the network. Is it possible for us to add a catch-all mock that throws an error, something that would act similarly to nock's `disableNetConnect`?

Comment (mcmire) [reply tests/setupTestsAfterEnv.ts:14]: I misspoke: it *does* throw if a recording is missing. This assumes that nothing will catch that error, though. CollectiblesController has a fair amount of `try`/`catch`s, so what would have been errors were simply warnings.

Comment (Gudahtt) [reply tests/setupTestsAfterEnv.ts:14]: Oh... that's great! Well. I am for switching to Polly then. Seems equally effective and at least a bit easier to use than nock.

**Comment by mcmire** [comment]
Okay, so it sounds like the next step here is to try to make some changes to the structure of these tests. That should mitigate some of the pain points with Nock I mentioned. 

I would also like to see if I can modify Nock so that when there is a request made that you haven't mocked yet, it will output information about that request in a similar way that Polly does. That should make working with Nock easier as it will be easy to understand which request was made and how exactly you can mock that request.

**Comment by mcmire** [comment]
I ran into more issues with Nock today and wanted to write them down. I was trying to write some tests in the extension and was trying to figure out why, despite my attempts to mock requests to Infura, Nock was producing an error indicating that a request was made but it couldn't find a mock matching the request. I eventually realized a few things:

1. Every time you say `.get` or `.post` you're adding an interceptor. Once a request is made, if that request matches an interceptor, that interceptor is removed from a list. That means that if two requests are made with identical request bodies, then you need to say `.post` twice. You can get around this by calling `.persist`, but then you have to call `nock.cleanAll()` in an `afterEach` or something to clear the "persisted" nocks. 
2. The error message that Nock produces is really confusing. It'll say "Nock: No match for request <some request> Got instead <some other request>". Which is the expected and which is the actual? Eventually, after staring at it for a bit, I realized the first is the actual and the second is the expected, but even so, the second request didn't contain a request body, even though I'd specified an expected request body in my test. 
3. In cases where you have a try/catch around a request and you don't throw the error appropriately, you won't get a proper error message when running the test. You can get around this by assigning the Nock mock to `scope` and then calling `scope.done()` at the end of the test, except that what happens if the test throws an error before `scope.done()`? So you have to end up commenting out some things in your test that would cause the error so that `scope.done()` is run appropriately. Really awkward.

**Comment by mcmire** [comment]
More Nock fun today. It appears that if you call `nock.cleanAll()` in an `afterEach` (which is necessary if you're using the `.persist()` option for a nock) and you have tests which add nocks, and your implementation makes multiple requests, then only the first test that has nocks will work; the rest won't. For instance:

``` javascript
describe('Test case', () => {
  afterEach(() => {
    nock.cleanAll();
  });

  it('does something', () => {
    // this will work
    nock('http://google.com')
      .post('/foo')
      .reply(200);
    
    callFunctionUnderTest();
  });

  it('does something else', () => {
    // this won't
    nock('http://google.com')
      .post('/foo')
      .reply(200);
    
    callFunctionUnderTest();
  });
});
```

Maybe the solution here is *not* to use `.persist()`, ever. That's fine, but people now need to learn this, and that `nock.cleanAll()` is a footgun.
